File: FruitWholesale.Client/src/app/features/collection/collection.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { Collection, SaveCollection } from '../../core/models/transactions.model';
import { toHttpParams } from '../../core/utils/http-params.util';

export interface PendingSettlementPreview {
  shopID: number;
  shopName?: string;
  pendingCount: number;
  pendingTotal: number;
}

export interface SettlementResult {
  settlementID: number;
  shopID: number;
  settlementDate: string;
  totalAmount: number;
  pendingCount: number;
}

@Injectable({ providedIn: 'root' })
export class CollectionService {
  private readonly baseUrl = `${environment.apiUrl}/collection`;

  constructor(private readonly http: HttpClient) {}

  getPaged(request: PaginationRequest, shopId?: number | null, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<Collection>>(this.baseUrl, {
      params: toHttpParams({ ...request, shopId, fromDate, toDate })
    });
  }

  create(dto: SaveCollection): Observable<Collection> {
    return this.http.post<Collection>(this.baseUrl, dto);
  }

  update(id: number, dto: SaveCollection): Observable<Collection> {
    return this.http.put<Collection>(`${this.baseUrl}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  getPendingSettlementPreview(shopId: number): Observable<PendingSettlementPreview> {
    return this.http.get<PendingSettlementPreview>(`${this.baseUrl}/pending-settle/preview`, {
      params: toHttpParams({ shopId })
    });
  }

  settle(dto: { shopID: number; settlementDate: string }): Observable<SettlementResult> {
    return this.http.post<SettlementResult>(`${this.baseUrl}/settle`, dto);
  }
}

File: FruitWholesale.Mobile/lib/features/collections/collection_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'collection_models.dart';

class CollectionService {
  final ApiClient _api;

  CollectionService(this._api);

  Future<PaginatedList<Collection>> getPaged({
    int pageNumber = 1,
    int pageSize = 20,
    int? shopId,
    String? fromDate,
    String? toDate,
  }) async {
    final json = await _api.get('/collection', query: {
      'pageNumber': pageNumber,
      'pageSize': pageSize,
      if (shopId != null) 'shopId': shopId,
      if (fromDate != null) 'fromDate': fromDate,
      if (toDate != null) 'toDate': toDate,
    });
    return PaginatedList.fromJson(json as Map<String, dynamic>, Collection.fromJson);
  }

  Future<Collection> getById(int id) async {
    final json = await _api.get('/collection/$id');
    return Collection.fromJson(json as Map<String, dynamic>);
  }

  Future<Collection> create(Collection collection) async {
    final json = await _api.post('/collection', body: collection.toSaveJson());
    return Collection.fromJson(json as Map<String, dynamic>);
  }

  Future<Collection> update(int id, Collection collection) async {
    final json = await _api.put('/collection/$id', body: collection.toSaveJson());
    return Collection.fromJson(json as Map<String, dynamic>);
  }

  Future<CollectionSettlementPreview> getPendingSettlementPreview(int shopId) async {
    final json = await _api.get('/collection/pending-settle/preview', query: {'shopId': shopId});
    return CollectionSettlementPreview.fromJson(json as Map<String, dynamic>);
  }

  Future<CollectionSettlementResult> settle({required int shopId, required DateTime settlementDate}) async {
    final json = await _api.post('/collection/settle', body: {
      'shopID': shopId,
      'settlementDate': settlementDate.toIso8601String(),
    });
    return CollectionSettlementResult.fromJson(json as Map<String, dynamic>);
  }
}
